flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaSinkITCase.java

```diff
@@ -0,0 +1,634 @@
+                LongStream.range(1, lastCheckpointedRecord.get().get() + 1)
+                        .boxed()
+                        .collect(Collectors.toList()));
+    }
+
+    @Test
+    public void testAbortTransactionsAfterScaleInBeforeFirstCheckpoint() throws Exception {
+        // Run a first job opening 5 transactions one per subtask and fail in async checkpoint phase
+        final Configuration config = new Configuration();
+        config.set(CoreOptions.DEFAULT_PARALLELISM, 5);
+        try {
+            executeWithMapper(new FailAsyncCheckpointMapper(0), config, null);
+        } catch (Exception e) {
+            assertThat(
+                    e.getCause().getCause().getMessage(),
+                    containsString("Exceeded checkpoint tolerable failure"));
+        }
+        assertTrue(deserializeValues(drainAllRecordsFromTopic(topic)).isEmpty());
+
+        // Second job aborts all transactions from previous runs with higher parallelism
+        config.set(CoreOptions.DEFAULT_PARALLELISM, 1);
+        failed.get().set(true);
+        executeWithMapper(
+                new FailingCheckpointMapper(failed, lastCheckpointedRecord), config, null);
+        final List<ConsumerRecord<byte[], byte[]>> collectedRecords =
+                drainAllRecordsFromTopic(topic);
+        assertEquals(
+                deserializeValues(collectedRecords),
+                LongStream.range(1, lastCheckpointedRecord.get().get() + 1)
+                        .boxed()
+                        .collect(Collectors.toList()));
+    }
+
+    private void executeWithMapper(
+            MapFunction<Long, Long> mapper,
+            Configuration config,
+            @Nullable String transactionalIdPrefix)
+            throws Exception {
+        final StreamExecutionEnvironment env = new LocalStreamEnvironment(config);
+        env.enableCheckpointing(100L);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        final DataStreamSource<Long> source = env.fromSequence(1, 10);
+        final DataStream<Long> stream = source.map(mapper);
+        final KafkaSinkBuilder<Long> builder =
+                new KafkaSinkBuilder<Long>()
+                        .setKafkaProducerConfig(getKafkaClientConfiguration())
+                        .setDeliverGuarantee(DeliveryGuarantee.EXACTLY_ONCE)
+                        .setBootstrapServers(KAFKA_CONTAINER.getBootstrapServers())
+                        .setRecordSerializer(new RecordSerializer(topic));
+        if (transactionalIdPrefix == null) {
+            transactionalIdPrefix = "kafka-sink";
+        }
+        builder.setTransactionalIdPrefix(transactionalIdPrefix);
+        stream.sinkTo(builder.build());
+        env.execute();
+        checkProducerLeak();
+    }
+
+    private void testRecoveryWithAssertion(
+            DeliveryGuarantee guarantee, java.util.function.Consumer<List<Long>> recordsAssertion)
+            throws Exception {
```

Comment:
FYI, there is another pattern that can be used to implement such a wrapping setup/cleanup code.
```
   AutoCloseableResult testRecovery(DeliveryGuarantee guarantee) {
       // execute common code
      result = // fetch result
      AutoCloseable after =  () -> { // after code };
      return wrap(result, after);
   }
```   
You can then use the return value in a try-and-resource and add all your assertions in the block. It has huge benefits over your pattern when you have checked exceptions and often is easier on the eye when the auto-formatter went over it.

Reply:
Thanks for the suggestion, I get the idea. I will try to facilitate it next time or do you want to have the tests refactored?